python/glow/gwas/log_reg.py

```diff
@@ -182,20 +198,47 @@ def _logistic_regression_inner(genotype_pdf: pd.DataFrame, log_reg_state: LogReg
     c, d: covariate
     '''
     X = np.column_stack(genotype_pdf[_VALUES_COLUMN_NAME].array)
+
+    # For approximate Firth correction, we perform a linear residualization
+    if correction == correction_approx_firth:
+        Q = np.linalg.qr(C)[0]
+        X = _residualize_in_place(X, Q)
+
     with oe.shared_intermediates():
         X_res = _logistic_residualize(X, C, Y_mask, log_reg_state.gamma, log_reg_state.inv_CtGammaC)
         num = gwas_fx._einsum('sgp,sp->pg', X_res, log_reg_state.Y_res)**2
         denom = gwas_fx._einsum('sgp,sgp,ps->pg', X_res, X_res, log_reg_state.gamma)
     t_values = np.ravel(num / denom)
     p_values = stats.chi2.sf(t_values, 1)
 
-    if fallback_method != fallback_none:
-        # TODO: Call approx firth here
-        ()
-
     del genotype_pdf[_VALUES_COLUMN_NAME]
     out_df = pd.concat([genotype_pdf] * log_reg_state.Y_res.shape[1])
     out_df['tvalue'] = list(np.ravel(t_values))
     out_df['pvalue'] = list(np.ravel(p_values))
     out_df['phenotype'] = phenotype_names.repeat(genotype_pdf.shape[0]).tolist()
+
+    if correction != correction_none:
+        correction_indices = list(np.where(out_df['pvalue'] < pvalue_threshold)[0])
+        if correction == correction_approx_firth:
+            out_df['effect'] = np.nan
+            out_df['stderr'] = np.nan
+            for correction_idx in correction_indices:
+                snp_idx = correction_idx % X.shape[1]
+                pheno_idx = int(correction_idx / X.shape[1])
+                approx_firth_snp_fit = correct_approx_firth(
+                    X[:, snp_idx],
+                    Y[:, pheno_idx],
+                    log_reg_state.approx_firth_state.logit_offset[:, pheno_idx],
+                    Y_mask[:, pheno_idx]
+                )
+                if approx_firth_snp_fit is not None:
+                    out_df.effect.iloc[correction_idx] = approx_firth_snp_fit.effect
+                    out_df.stderr.iloc[correction_idx] = approx_firth_snp_fit.stderr
+                    out_df.tvalue.iloc[correction_idx] = approx_firth_snp_fit.tvalue
+                    out_df.pvalue.iloc[correction_idx] = approx_firth_snp_fit.pvalue
+                else:
+                    print(f"Could not correct {out_df.iloc[correction_idx]}")
```

Comment:
This is not really helpful. We should put something in the result that indicates that the correction failed.